streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1848,113 +1907,101 @@ object ZChannel {
     f: (OutDone, OutDone) => OutDone
   )(implicit trace: Trace): ZChannel[Env, InErr, InElem, InDone, OutErr, OutElem, OutDone] =
     unwrapScopedWith { scope =>
-      {
-        for {
-          input         <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
-          queueReader    = ZChannel.fromInput(input)
-          n             <- ZIO.succeed(n)
-          bufferSize    <- ZIO.succeed(bufferSize)
-          mergeStrategy <- ZIO.succeed(mergeStrategy)
-          queue         <- Queue.bounded[ZIO[Env, OutErr, Either[OutDone, OutElem]]](bufferSize)
-          _             <- scope.addFinalizer(queue.shutdown)
-          cancelers     <- Queue.unbounded[Promise[Nothing, Unit]]
-          _             <- scope.addFinalizer(cancelers.shutdown)
-          lastDone      <- Ref.make[Option[OutDone]](None)
-          errorSignal   <- Promise.make[Nothing, Unit]
-          permits       <- Semaphore.make(n.toLong)
-          pull          <- (queueReader >>> channels).toPullIn(scope)
-          evaluatePull = (pull: ZIO[Env, OutErr, Either[OutDone, OutElem]]) =>
-                           pull.flatMap {
-                             case Left(done) =>
-                               ZIO.succeed(Some(done))
-                             case Right(outElem) =>
-                               queue
-                                 .offer(ZIO.succeed(Right(outElem)))
-                                 .as(None)
-                           }
-                             .repeatUntil(_.isDefined)
-                             .flatMap {
-                               case Some(outDone) =>
+      for {
+        input         <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
+        incoming       = ZChannel.fromInput(input)
+        n             <- ZIO.succeed(n)
+        bufferSize    <- ZIO.succeed(bufferSize)
+        mergeStrategy <- ZIO.succeed(mergeStrategy)
+        outgoing      <- Queue.bounded[ZIO[Env, Either[OutErr, OutDone], OutElem]](bufferSize)
+        _             <- scope.addFinalizer(outgoing.shutdown)
+        cancelers     <- Queue.unbounded[Promise[Nothing, Unit]]
+        _             <- scope.addFinalizer(cancelers.shutdown)
+        lastDone      <- Ref.make[Option[OutDone]](None)
+        errorSignal   <- Promise.make[Nothing, Unit]
+        permits       <- Semaphore.make(n.toLong)
```

Comment:
I didn't pay much attention to this to be honest as it's executed only at the creation of the channel but yeah it might have a cost for something like zio-grpc.